staging/src/k8s.io/node-api/pkg/apis/node/v1alpha1/types.go

```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// RuntimeClass defines a class of runtime supported in the cluster.
+type RuntimeClass struct {
+	metav1.TypeMeta `json:",inline"`
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Specification of the RuntimeClass
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	Spec RuntimeClassSpec `json:"spec" protobuf:"bytes,2,name=spec"`
+}
+
+// RuntimeClassSpec is a specification of a RuntimeClass.
+type RuntimeClassSpec struct {
+	// RuntimeHandler specifies the underlying runtime the CRI calls to handle pod and/or container
+	// creation. The possible values are specific to a given configuration & CRI implementation.
```

Comment:
Are we speccing the format of this string?  e.g. containerd.io/containerd or cri-o.io/cri-o ?

Reply:
For now I'm using the same validation as for the ObjectMeta name (DNS 1123 subdomain). We might want to expand that to allow paths later, but it's easier to start with tighter validation.  I'll add this to the comment.

I don't think there's any reason to namespace the runtime handler. This field is only ever interpreted by the runtime, and the expectation is that the admin set it up to map to a specific config that the runtime understands. If we prefixed it with `containerd.io`, it would be up to containerd to then strip out that prefix (and I guess reject any request that didn't have that prefix).

Reply:
IIUC, what you're getting at is discovery? I.e. once we have scheduling support for RuntimeClass, how do nodes broadcast the runtimes they support, and how does the the cluster match that to a RuntimeClass.

I think it's an open question of whether nodes would "broadcast" supported RuntimeHandlers, or RuntimeClasses (or whether discovery should be supported at all).

Even if nodes did broadcast supported RuntimeHandlers, I think it's debateable whether 2 different CRI implementations configured with the same RuntimeHandler should be treated separately, or as 2 valid implementations of the same handler.

Given these open questions, I'd prefer to punt on this until we actually have a design for this feature.

Reply:
We spoke offline.  here's my proposed comments:

```
RuntimeHandler specifies the underlying runtime and configuration that the CRI implementation will use to handle pods of this class. The possible values are specific to the node configuration & CRI implementation.  It is assumed that all handlers are available and equivalent on every node.

If this is not specified, a default will be chosen by the implementation.